### PR TITLE
fix expand op, add tensor and parameter inferencefor support for argument expand_times

### PR DESCRIPTION
**Pull request by liym27**
#### 1. Support  argument `expand_times` is a tensor variable.
**Before**：
`expand_times` support tensor, but the tensor must be a element of `expand_times` list;
`layers.expand(x, expand_times=[times_1, 3])`
> `expand_times` is a list,  `times_1` can be  a tensor variable.  

**After**：
Also support that `expand_times` is a tensor variable.
`layers.expand(x, expand_times=expand_times)`
> `expand_times` can be tensor varible.
#### 2. Support inferring the dimensions corresponding to constants when `expand_times` is a list containing integer and tensor variable.

### CODE REVIEW

File: paddle/fluid/operators/expand_op.cc [hunk 1]
```diff
@@ -49,11 +50,15 @@ class ExpandOp : public framework::OperatorWithKernel {
...
+        PADDLE_ENFORCE_GT(
+            expand_times[i], 0,
+            "The element of Attr(expand_times) must greater than 0.");
         out_shape[i] = x_dims[i] * expand_times[i];
       }
     }
 
-    ctx->SetOutputDim("Out", framework::make_ddim(out_shape));
+    auto out_dims = framework::make_ddim(out_shape);
+    ctx->SetOutputDim("Out", out_dims);
```
Comment (lanxianghit) [inline paddle/fluid/operators/expand_op.cc]: 拆开写的目的是？

File: paddle/fluid/operators/expand_op.cc [hunk 2]
```diff
@@ -83,7 +88,14 @@ class ExpandOpMaker : public framework::OpProtoAndCheckerMaker {
...
              "X is the input to be expanded.");
-    AddInput("expand_times_tensor", "(Tensor Tensor<int>), epxand times for X")
+    AddInput("ExpandTimes",
+             "(Tensor<int>), optional). If provided, expand according to "
+             "this given expand times. It has a higher priority than "
+             "expand_times_tensor.")
+        .AsDispensable();
+    AddInput("expand_times_tensor",
+             "(Tensor Tensor<int>), epxand times for X."
+             "It has a higher priority than expand_times.")
```
Comment (lanxianghit) [inline paddle/fluid/operators/expand_op.cc]: > "It has a higher priority than expand_times.")
到底谁的优先级高？

Comment (liym27) [reply paddle/fluid/operators/expand_op.cc]: 优先级由高到低，依次是：ExpandTimes  > expand_times_tensor > expand_times，
即 Tensor > list (contain tensor) > list (only contain integer )

Comment (liym27) [reply paddle/fluid/operators/expand_op.cc]: 优先级由高到低，依次是：ExpandTimes > expand_times_tensor > expand_times，
即 Tensor > list (contain tensor) > list (only contain integer )

File: paddle/fluid/operators/expand_op.cc (hunk 1 above)
Comment (liym27) [reply paddle/fluid/operators/expand_op.cc]: 之前拆开写是为了方便打印，已改。

File: paddle/fluid/operators/expand_op.cc (hunk 2 above)
Comment (liym27) [reply paddle/fluid/operators/expand_op.cc]: 已经增加了说明。

Comment (liym27) [reply paddle/fluid/operators/expand_op.cc]: 已经增加了说明。

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -9902,54 +9902,83 @@ def expand(x, expand_times, name=None):
...
+                temp_out = helper.create_variable_for_type_inference('int32')
+                fill_constant([1], 'int32', ele, force_cpu=True, out=temp_out)
+                new_expand_times_tensor.append(temp_out)
+        return new_expand_times_tensor
 
     if in_dygraph_mode():
         inputs = {'X': x}
         attrs = {'expand_times': expand_times}
     else:
+        if isinstance(expand_times, Variable):
```
Comment (phlrain) [inline python/paddle/fluid/layers/nn.py:10356]: test_layers.py 中新增这个op的测试，增加代码覆盖率

**Comment by jiangjiajun** [comment]
重新测试了接口，使用无问题，且满足目前的需求。

**Review by lanxianghit — DISMISSED** [review]
LGTM

**Review by lanxianghit — DISMISSED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by heavengate — APPROVED** [review]

**Pull request merged**